edb/lib/ext/auth.edgeql

```diff
@@ -17,4 +17,49 @@
 #
 
 
-CREATE EXTENSION PACKAGE auth VERSION '1.0';
+CREATE EXTENSION PACKAGE auth VERSION '1.0' {
+    set ext_module := "ext::auth";
+
+    create module ext::auth;
+
+    create type ext::auth::Identity {
+        create required property provider: std::str;
+
+        # Standard OpenID claims
+        create required property sub: std::str;
+        create property name: std::str;
+        create property given_name: std::str;
+        create property family_name: std::str;
+        create property middle_name: std::str;
+        create property nickname: std::str;
+        create property preferred_username: std::str;
+        create property profile: std::str;
+        create property picture: std::str;
+        create property website: std::str;
+        create property gender: std::str;
+        create property birthdate: std::str;
+        create property zoneinfo: std::str;
+        create property locale: std::str;
+        create property updated_at: std::datetime;
```

Comment:
I usually keep only `sub` here (indexed with provider? unique?) and make the rest computed properties of a single JSON "claims" property, so that it's easier to update as a whole and allow storing custom claims (also semantically allows non-OpenID user profiles)

Reply:
Yes to the constraint :v:

> We would store a `property claims: std::json`, right? And then use `json_get` to map from the json claim blob to the OpenID-like property name, right?

Yes!

> How would we deal with the fact that each provider might use different names for something like `picture` (`avatar_url` for GitHub, for example)?

Theoretically, we can choose either preprocessing or postprocessing - your current code would preprocess the received claims and store normalized properties, right? We can probably do the same with an additional `normalized_claims` property. (Yes I tend to keep the original claims we received, even just for debugging purposes)

Or, `picture` can be a smart computed property that also considers the `provider` value, like `picture := claims.avatar_url if .provider = 'github' else claims.picture` ... but I'm not so sure if we want to go there, because every time we add a provider, we may have to change the schema.

Reply:
> Theoretically, we can choose either preprocessing or postprocessing - your current code would preprocess the received claims and store normalized properties, right?

Yeah, I felt like that has the best DX for the consumer: we just document these well-known OpenID-like properties and the application developer doesn't have to think about the provider shape (and if it changes in the future, _we_ have the burden of updating our mappings).

> We can probably do the same with an additional normalized_claims property.

Why store this normalized as JSON rather than properties?

> (Yes I tend to keep the original claims we received, even just for debugging purposes)

Cool.

One point of clarification here: most of these providers are _not_ OpenID providers, so the data might come from disparate endpoints from the resource server. In GitHub's case, the `/user` and `/user/emails` endpoints map pretty cleanly to our own types, but other providers might require multiple requests for different important identity data. Given this, do you think it makes sense to maybe make the "original claims" JSON property hold some kind of map from URI -> JSON response?

```json 
{
  "https://api.github.com/user": {
    "id": "abc123",
    "login": "octocat"
  },
  "https://api.github.com/user/emails": [
    {
      "email": "octocat@example.com",
      "verified": true,
      "primary": true
    }
  ]
}
```

Reply:
> Why store this normalized data as JSON rather than properties?

Good point! If we keep the original `claims` already, we don't need another cache layer before the individual properties.

> Given this, do you think it makes sense to maybe make the "original claims" JSON property hold some kind of map from URI -> JSON response?

Oh wow, now it's getting more complicated than I expected. That makes sense, however: If we are feeling relaxed with modifying the schema when adding providers, I'm thinking maybe we can have `GithubIdentity extending Identity` with extra properties for original responses (for easier and consistent user API access), together with computed properties like `picture := .user_response.avatar_url`. (Or rather, `Identity` holds the preprocessed normalized properties as the "cache")

Reply:
Yeah, my original idea was to have different types for each provider, but in our initial review of my proposed schema we (I think @elprans maybe suggested it?) came to the conclusion that there was more value in providing the application developer with a single consistent interface and not much value (at least for the application developer) to having a bunch of subtypes that captures the original shape of the data, since it might change over time.

I still see value in capturing raw data for debugging purposes though, so if we can come to some reasonable design there that doesn't add a lot of noise to the server code here, I'm all for it!